include/dlaf/matrix/panel.h

```diff
@@ -0,0 +1,450 @@
+
+namespace dlaf {
+namespace matrix {
+
+/// Panel (1D array of tiles)
+///
+/// 1D array of tiles, i.e. a Row or Column panel strictly related to a given dlaf::Matrix (from the
+/// coords point of view)
+template <Coord axis, class T, Device device>
+struct Panel;
+
+template <Coord axis, class T, Device device>
+struct Panel<axis, const T, device> : protected Matrix<T, device> {
+  // Note:
+  // This specialization acts as base for the RW version of the panel,
+  // moreover allows the casting between references (i.e. Panel<const T>& = Panel<T>)
+
+  using TileType = Tile<T, device>;
+  using ConstTileType = Tile<const T, device>;
+  using BaseT = Matrix<T, device>;
+
+  Panel(Panel&&) = default;
+
+  /// On destruction, reset the panel
+  ///
+  /// Resetting the panel implies removing external dependencies
+  virtual ~Panel() noexcept {
+    reset();
+  }
+
+  /// Return an IterableRange2D with a range over all tiles of the panel (considering the offset)
+  auto iterator() const noexcept {
+    return range_;
+  }
+
+  /// Return the rank which this (local) panel belongs to
+  auto rankIndex() const noexcept {
+    return dist_matrix_.rankIndex();
+  }
+
+  /// Return the Distribution of the parent matrix
+  auto parentDistribution() const noexcept {
+    return dist_matrix_;
+  }
+
+  /// Set a specific index to point to the specified external tile
+  ///
+  /// It is possible to set to an external tile on an index if, since last reset() or from
+  /// the creation of the panel, the specific index:
+  /// - has not been accessed, neither on read or read/write
+  /// - has not been already set to an external tile
+  ///
+  /// @pre @p index must be a valid index for the current panel size
+  void setTile(const LocalTileIndex& index, hpx::shared_future<ConstTileType> new_tile_fut) {
+    DLAF_ASSERT(index.isIn(dist_matrix_.localNrTiles()), index, dist_matrix_.localNrTiles());
+    DLAF_ASSERT(internal_.count(linearIndex(index)) == 0, "internal tile have been already used", index);
+    DLAF_ASSERT(!isExternal(index), "already set to external", index);
+
+    external_[linearIndex(index)] = std::move(new_tile_fut);
+  }
+
+  /// Access a Tile of the panel in read-only mode
+  ///
+  /// This method is very similar to the one available in dlaf::Matrix.
+  ///
+  /// @p index is in the coordinate system of the matrix which this panel is related to
+  hpx::shared_future<ConstTileType> read(const LocalTileIndex& index) {
+    DLAF_ASSERT_HEAVY(index.isIn(dist_matrix_.localNrTiles()), index, dist_matrix_.localNrTiles());
+
+    const SizeType internal_linear_idx = linearIndex(index);
+    if (isExternal(index)) {
+      return external_[internal_linear_idx];
+    }
+    else {
+      internal_.insert(internal_linear_idx);
+      return BaseT::read(fullIndex(index));
+    }
+  }
+
+  /// Set the panel to a new offset (with respect to the "parent" matrix)
+  ///
+  /// @pre offset cannot be less than the offset has been specifed on construction
```

Comment:
It's a good point. I think we can easily deal with this change:
- using two different member functions
- using an additional parameter (being it a boolean, a tag, ...)

A remark: the coordinate system in the panel is, as it has been implemented, always equal to the one of the local parent matrix. This means that no matter the hook point for the shrinking, the tiles will always be accessed using LocalTileIndex of the parent matrix.

See the next example, where a panel `P` with its parent matrix `M` (just local part) is depicted togheter with row indices to be used for accessing the tiles.
```
# full panel			# shrink from top		# shrink from bottom
0 P  M M M M M M		  P  M M M M M M		0 P  M M M M M M
1 P  M M M M M M		  P  M M M M M M		1 P  M M M M M M
2 P  M M M M M M		2 P  M M M M M M		2 P  M M M M M M
3 P  M M M M M M		3 P  M M M M M M		3 P  M M M M M M
4 P  M M M M M M		4 P  M M M M M M		  P  M M M M M M
5 P  M M M M M M		5 P  M M M M M M		  P  M M M M M M
```

Reply:
I understand that indexing is coherent, however I think that you need to know the correct size for the bcast in the shrink from bottom case, otherwise all tiles will be communicated.

Reply:
>     * using two different member functions
>     * using an additional parameter (being it a boolean, a tag, ...)

Maybe a single member function that takes both offset and size?

Reply:
It looks reasonable, I'm working on it.

Considering the use-case you have in mind, do you think it is enough to allow the shrinking from both edges just once initialized, or do you want also to be able to build a panel that allocates just a (start, end) range of tiles?

Currently it is possible to build a panel that does not begin with the beginning of the matrix, but the end is fixed to the last tile of the parent matrix.

Reply:
I don't see the use case for this, therefore I would not.